include/downloader_pool.h

```diff
@@ -0,0 +1,248 @@
+#ifndef DOWNLOADER_POOL_H
+#define DOWNLOADER_POOL_H
+
+#include <optional>
+#include <thread>
+#include <functional>
+#include <map>
+#include <queue>
+#include <mutex>
+#include <vector>
+#include <list>
+
+#include "sockets.h"
+#include "http_downloader.h"
+#include "response.h"
+#include "uri.h"
+
+namespace Downloaders::Concurrency
+{
+	/// Callback that should accept HTTP response and should specify if download should be repeated
+	using FDownloadCallback = 
+		std::function<bool(std::optional<HTTP::CHTTPResponse>&&)>;
+
+	/// Special thread pool that carries conccurent HTTP download
+	/// provided by SocketClass that should inherit CSocket class
+	/// After each download from some URI running callback function
+	/// with passed response
+	template<class SocketClass>
+	class CConcurrentDownloader
+	{
+	static_assert(std::is_base_of<Sockets::CSocket, SocketClass>(), "SocketClass should inherit CSocket");
+	private:
+		// Because of language limitations I cannot put this in the end of class
+		// For better clarity using simple struct instead of std::tuple
+		struct WorkingTuple
+		{
+			Downloaders::CHTTPDownloader<SocketClass> m_downloader;
+			CURI m_URIToDownload;
+			std::thread m_runningThread;
+			FDownloadCallback m_fCallback;
+
+			WorkingTuple(const CURI &newURI) : m_URIToDownload{newURI}
+			{
+
+			}
+		};
```

Comment:
Looks good. But if you call it tuple. Could the std::tuple be used for this?

Reply:
Removed and replaced types with std::tuple and std::pair